src/libcollections/btree/map.rs

```diff
@@ -529,6 +537,174 @@ impl<K: Ord, V> BTreeMap<K, V> {
         }
     }
 
+    /// Moves all elements from `other` into `Self`, leaving `other` empty.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::collections::BTreeMap;
+    ///
+    /// let mut a = BTreeMap::new();
+    /// a.insert(1, "a");
+    /// a.insert(2, "b");
+    /// a.insert(3, "c");
+    ///
+    /// let mut b = BTreeMap::new();
+    /// b.insert(3, "d");
+    /// b.insert(4, "e");
+    /// b.insert(5, "f");
+    ///
+    /// a.append(&mut b);
+    ///
+    /// assert_eq!(a.len(), 5);
+    /// assert_eq!(b.len(), 0);
+    ///
+    /// assert_eq!(a[&1], "a");
+    /// assert_eq!(a[&2], "b");
+    /// assert_eq!(a[&3], "d");
+    /// assert_eq!(a[&4], "e");
+    /// assert_eq!(a[&5], "f");
+    /// ```
+    #[stable(feature = "append", since = "1.4.0")]
+    pub fn append(&mut self, other: &mut Self) {
+        // Do we have to append anything at all?
+        if other.len() == 0 {
+            return;
+        }
+
+        // We can just swap `self` and `other` if `self` is empty.
+        if self.len() == 0 {
+            mem::swap(self, other);
+            return;
+        }
+
+        // First, we merge `self` and `other` into a sorted sequence in linear time.
+        let mut self_iter = mem::replace(self, BTreeMap::new()).into_iter();
+        let mut other_iter = mem::replace(other, BTreeMap::new()).into_iter();
+        let self_cur = self_iter.next();
+        let other_cur = other_iter.next();
+        let iter = MergeIter {
+            left: self_iter,
+            right: other_iter,
+            left_cur: self_cur,
+            right_cur: other_cur,
+        };
+
+        // Second, we build a tree from the sorted sequence in linear time.
+        let mut length = 0;
+        let mut root = node::Root::new_leaf();
+
+        {
+            let mut cur_node = root.as_mut();
+
+            // Iterate through all key-value pairs, pushing them into nodes at the right level.
+            for (key, value) in iter {
+                match cur_node.force() {
+                    Leaf(mut leaf) => {
+                        // We are at a leaf, try to push key-value pair if it fits.
+                        if leaf.len() < node::CAPACITY {
+                            leaf.push(key, value);
+                            cur_node = leaf.forget_type();
+                        } else {
+                            // No space left, go up and push there.
+                            cur_node = leaf.forget_type();
```

Comment:
This can be moved after `loop {`, so that `parent.forget_type()` in the Ok/else branch becomes redundant.